### PR TITLE
The Defenestration of recreate's bloat

### PR DESCRIPTION
**Pull request by enkore**

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
seems way simpler now. \o/

review of this sucks a bit due to github's bad diff display that does not show minimal diff (esp. for indentation changes).

**Comment by RonnyPfannschmidt** [comment]
lovely change

File: src/borg/testsuite/archiver.py [hunk 1]
```diff
@@ -1626,6 +1629,40 @@ def raise_eof(*args):
...
+        self.cmd('check', self.repository_location)
+        # Then check that the cache on disk matches exactly what's in the repo.
+        with self.open_repository() as repository:
+            manifest, key = Manifest.load(repository)
+            with Cache(repository, key, manifest, sync=False) as cache:
+                original_chunks = cache.chunks
+            cache.destroy(repository)
+            with Cache(repository, key, manifest) as cache:
+                correct_chunks = cache.chunks
+        assert original_chunks != correct_chunks
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: huh? why unequal?

File: src/borg/testsuite/archiver.py [hunk 2]
```diff
@@ -1626,6 +1629,40 @@ def raise_eof(*args):
...
+        seen = set()
+        for id, (refcount, size, csize) in correct_chunks.iteritems():
+            o_refcount, o_size, o_csize = original_chunks[id]
+            assert refcount == o_refcount
+            assert size == o_size
+            assert csize == o_csize
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: use tuple comparison to do all at once? assuming that py.test will show the complete tuple so one still sees what's wrong.

File: src/borg/archive.py [hunk 3]
```diff
@@ -1476,127 +1468,82 @@ def item_is_hardlink_master(item):
...
+            return self.cache.chunk_incref(chunk_id, target.stats)
+
+        compression_spec, chunk = self.key.compression_decider2.decide(chunk)
+        overwrite = self.recompress
+        if self.recompress and not self.always_recompress and chunk_id in self.cache.chunks:
+            # Check if this chunk is already compressed the way we want it
+            old_chunk = self.key.decrypt(None, self.repository.get(chunk_id), decompress=False)
+            if Compressor.detect(old_chunk.data).name == compression_spec['name']:
+                # Stored chunk has the same compression we wanted
+                overwrite = False
```
Comment (enkore) [inline src/borg/archive.py]: > --always-recompress
>    always recompress chunks, don’t skip chunks already compressed with the samealgorithm. [sic]

File: src/borg/testsuite/archiver.py (hunk 1 above)
Comment (enkore) [inline src/borg/testsuite/archiver.py]: To explicate that these are different objects by way of the sync + reopened cache.

File: src/borg/testsuite/archiver.py (hunk 2 above)
Comment (RonnyPfannschmidt) [inline src/borg/testsuite/archiver.py]: potentially the import was triggered before the pytest assertion rewriter got active

one can force that by some invocations in the conftest

File: src/borg/archive.py (hunk 3 above)
Comment (ThomasWaldmann) [inline src/borg/archive.py]: ok, pity.

File: src/borg/testsuite/archiver.py (hunk 1 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: yes, if that is the intent, please use "is not". != is a value comparison.

File: src/borg/testsuite/archiver.py (hunk 2 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: I've also sometimes missed details. @RonnyPfannschmidt can you help investigate, that sucks.

**Comment by haffenloher** [comment]
@ThomasWaldmann

> review of this sucks a bit due to github's bad diff display that does not show minimal diff (esp. for indentation changes).

I know I'm late but this may help for future reviews: Append `?w=1` to a github URL to view diffs without whitespace changes, like this:

https://github.com/borgbackup/borg/pull/1862/files?w=1

**Comment by ThomasWaldmann** [comment]
@haffenloher thanks, I'll try it next time.

**Pull request merged**